### ISSUE TITLE
[SDL3] X11 XGenericEventCookie compilation error

### ISSUE THREAD

**Issue by geugenm** [libsdl-org/SDL#7782]
The header with error: [`src/video/x11/SDL_x11xinput2.h`](https://github.com/libsdl-org/SDL/blob/main/src/video/x11/SDL_x11xinput2.h#L30), line 30

The compilation error: `conflicting types for ‘XGenericEventCookie’; have ‘struct XGenericEventCookie’`

### System information:
Linux-x86-64: kernel `6.3.6-1`
NVIDIA driver version: `530.41.03` (tried both on open-source and proprietary drivers)
NVML version: `12.530.41.03`

### Description
The compilation error is due to a conflicting type definition for XGenericEventCookie. In the src/video/x11/SDL_x11xinput2.h header file, XGenericEventCookie is defined as both a struct and a typedef. The correct way to define a struct is using the struct keyword, not with a typedef.

### How to solve
Just remove `typedef struct XGenericEventCookie XGenericEventCookie;` (30th line)

**Comment by geugenm** [comment]
I've tested it on WSL (ubuntu 22) on another machine with XMing. The error persists, the solution is the same. So, its a problem with X11, not just with my Manjaro KDE

**Comment by slouken** [comment]
I just built on WSL (ubuntu 22) with no issues.
I edited /etc/apt/sources.list, uncommended the deb-src lines, ran `sudo apt update` and `sudo apt build-dep libsdl2` and then ran cmake normally and it built just fine.

I'm trying to reproduce your environment. What is defined in the X11 section of your SDL_build_config.h?
Mine is:
```
#define SDL_VIDEO_DRIVER_X11 1
#define SDL_VIDEO_DRIVER_X11_DYNAMIC "libX11.so.6"
#define SDL_VIDEO_DRIVER_X11_DYNAMIC_XCURSOR "libXcursor.so.1"
#define SDL_VIDEO_DRIVER_X11_DYNAMIC_XEXT "libXext.so.6"
#define SDL_VIDEO_DRIVER_X11_DYNAMIC_XFIXES "libXfixes.so.3"
#define SDL_VIDEO_DRIVER_X11_DYNAMIC_XINPUT2 "libXi.so.6"
#define SDL_VIDEO_DRIVER_X11_DYNAMIC_XRANDR "libXrandr.so.2"
#define SDL_VIDEO_DRIVER_X11_DYNAMIC_XSS "libXss.so.1"
#define SDL_VIDEO_DRIVER_X11_HAS_XKBKEYCODETOKEYSYM 1
#define SDL_VIDEO_DRIVER_X11_SUPPORTS_GENERIC_EVENTS 1
#define SDL_VIDEO_DRIVER_X11_XCURSOR 1
#define SDL_VIDEO_DRIVER_X11_XDBE 1
#define SDL_VIDEO_DRIVER_X11_XFIXES 1
#define SDL_VIDEO_DRIVER_X11_XINPUT2 1
#define SDL_VIDEO_DRIVER_X11_XINPUT2_SUPPORTS_MULTITOUCH 1
#define SDL_VIDEO_DRIVER_X11_XRANDR 1
#define SDL_VIDEO_DRIVER_X11_XSCRNSAVER 1
#define SDL_VIDEO_DRIVER_X11_XSHAPE 1
```

**Comment by slouken** [comment]
@madebr, I was able to reproduce this. The problem is that somehow `SDL_VIDEO_DRIVER_X11_SUPPORTS_GENERIC_EVENTS` was not defined, even though `XGenericEventCookie` is available in Xlib.h

**Comment by madebr** [comment]
How did you reproduce it?
I have the same config as you've posted in https://github.com/libsdl-org/SDL/issues/7782#issuecomment-1576974211.

**Comment by madebr** [comment]
I can reproduce it when I configure with `-DHAVE_XGENERICEVENT=OFF`, to [force the "XGenericEvent" to fail](https://github.com/libsdl-org/SDL/blob/6150b5b3cbde0e592c4ffe822f66aa5f9c90c3d9/cmake/sdlchecks.cmake#L343-L355).

But why is it failing?

**Comment by geugenm** [comment]
## Here how I reproduce it:

1. Added latest [CPM](https://github.com/cpm-cmake/CPM.cmake) as a git submodule. Then I've created a cpm cmake script to compile sdl:


```
CPMAddPackage(
        NAME SDL3
        GITHUB_REPOSITORY libsdl-org/SDL
        GIT_TAG 6150b5b3cbde0e592c4ffe822f66aa5f9c90c3d9 # One of the latest commits
        OPTIONS
        
        "SDL_STATIC ON"
        "SDL_SHARED ON"

        "SDL_CCACHE ON"

        "SDL_VULKAN OFF"
        "SDL_ASSEMBLY OFF"
)
```



Then I've compiled the sdl using the preset:
```
      "name": "Debug", 
       "displayName": "Debug", 
       "description": "Debug build with assertions, debug info, and additional debugging options.", 
       "generator": "Ninja", 
       "binaryDir": "${sourceDir}/build/debug", 
       "cacheVariables": { 
         "CMAKE_BUILD_TYPE": "Debug", 
         "CMAKE_CXX_STANDARD": "20", 
         "CMAKE_CXX_FLAGS": "-Og -g -Wall -Wextra -Wpedantic -Wconversion -Wno-unused-function", 
         "CMAKE_C_FLAGS": "-Og -g -Wall -Wextra -Wpedantic -Wconversion -Wno-unused-function", 
         "CMAKE_CXX_FLAGS_RELEASE": "-fsanitize=address,undefined", 
         "CMAKE_C_FLAGS_RELEASE": "-fsanitize=address,undefined", 
         "CMAKE_EXPORT_COMPILE_COMMANDS": "ON"
```

## sdl_build_config
Here is my [sdl_build_config.h.intermediate](https://gist.github.com/geugenm/a3a72c486ba0738fbb888a1205839086)

**Comment by madebr** [comment]
Can you try https://github.com/libsdl-org/SDL/pull/7785?
Are the x11 headers installed on your system? If they are, where?
They can be found using `find / -iname "Xlib.h" 2>/dev/null`.

If my pr does not address your issue, can you please clear the build folder and run `cmake .. --trace-expand >log.txt 2>&1`?
Then please share  `build/log.txt`, `build/CMakeFiles/CMakeOutput.log`, `build/CMakeFiles/CMakeError.log` and `build/CMakeFiles/CMakeConfigureLog.yaml`  (not all of the last 3 files might be present)

**Comment by geugenm** [comment]
> It builds, but does it display any window?

Yep, it displayed it before. But now I don't need to change anything in sdl sources

**Comment by madebr** [comment]
The `Xlib.h`  in `/usr/include/X11/Xlib.h` is in a standard location.
Why doesn't `find_path` find it on your system?

What is the output of the following mini cmake project?
```cmake
cmake_minimum_required(VERSION 3.20)
project(x_headers)
find_path(X1 NAMES "X11/Xlib.h" PATHS /usr/include/X11)
message("X1: ${X1}")
find_path(X2 NAMES "X11/Xlib.h" PATHS /usr/include)
message("X2: ${X2}")
```
On my system, it is:
```
X1: /usr/include
X2: /usr/include
```

If it finds the header, can you break it by reproducing the SDL error?
Or  if it doesn't find the header, can you fix it?

Are you using a cmake toolchain?

**Comment by geugenm** [comment]
The output is the same:
```
-- X1: /usr/include
-- X2: /usr/include
```

I'm using default cmake + [CPM](https://github.com/cpm-cmake/CPM.cmake) which is actually extending the default `FetchContent` and then uses the cmakelists.txt inside the added library library. 

Now, I've removed and recompiled my project with the same commit (deleted all caches and build directory before) and the error does not occur anymore. Now everything works good. Done the same on WSL (where the error also presented) and the result is the same - no errors anymore. So now I can't reproduce the bug. I will try, maybe its the problem with cmake cache, but the last time (before this) I've removed build folders the error persisted.

**Comment by madebr** [comment]
Did you install any packages after your first SDL configuration?
If you installed X11 development libraries/headers after that, cmake will not perform another search.

**Comment by geugenm** [comment]
I have never added any libraries in all the time I have been using this operating system, only fonts and applications. I originally had all the XServer libraries (Manjaro KDE provided), nothing more was added to them. The last thing I've UPDATED which is connected to XServer was xorg-xinput (4 June).  But the error was before that. 

Firstly in the end of April I've compiled SDL and it was successfully compiled, I've used it without any problem. After that moment I was just UPDATING the existing packages (like xorg-xinput) via default Manjaro pamac. And in one moment I've got this compilation error. The same was on my laptop with WSL(XServer was also installed to it before). Even clean reclon of my repo with the project did not help in both situations. 

About my update policy. I'm updating when the notification about the available update pops up. So, maybe the problem is with those updates.
